### ISSUE TITLE
`fft` outputs non-contiguous

### ISSUE THREAD

**Issue by OverLordGoldDragon** [pytorch/pytorch#54376]
Unsure if intended; applies to contiguous inputs.

```python
import torch
import torch.fft

x =  torch.randn(200, 200)
print(torch.fft.fft(x, dim=0).is_contiguous(), x.is_contiguous())
```
```
>>> False True
```

cc @mruberry @peterbell10 @walterddr

**Comment by mruberry** [comment]
Thanks for reporting this, @OverLordGoldDragon.

@peterbell10, what are your thoughts? Should we produce contiguous outputs, or maybe update the documentation?

**Comment by OverLordGoldDragon** [comment]
@mruberry I'm fine with just documenting, the problem is the behavior's unexpected and caused trouble silently.

**Comment by peterbell10** [comment]
Yes, this is expected. For technical reasons, there is a significant performance cost to guaranteeing contiguous output. In many cases, the best we could do is just call `.contiguous()` on the tensor internally.

First speaking in general: mkl-fft and cuFFT don't support completely general tensors, e.g. only supporting a single batch dimension. For best performance though, we need to give all batched transforms to the FFT library in a single call. With those constraints, the transform output can't always be contiguous. So we either accept that, or add an explicit clone at the end to make it contiguous.

For the specific example given (`dim=0` transform), I believe I benchmarked that both cuFFT and mkl-fft perform significantly better when each individual transform in the batch is contiguous in the output. So since we're not guaranteeing contiguous output in general, it makes sense to me to prefer the layout that performs better.

**Comment by peterbell10** [comment]
Also, is it reasonable to expectat that every operator output is contiguous? It's certainly true that simple point-wise operations default to contiguous output. However, I would expect many of the linear algebra operators default to Fortran order, one specific example is [`torch.lu_solve`](https://pytorch.org/docs/master/generated/torch.lu_solve.html#torch-lu-solve) which doesn't have any disclaimer.

**Comment by OverLordGoldDragon** [comment]
> is it reasonable to expect that every operator output is contiguous?

This may be my mistake. I just figured contiguous-in -> contiguous-out, but if this doesn't hold of most ops (and they don't document it), guess there's nothing to do.

**Comment by SsnL** [comment]
I don't think that users should assume continuity of outputs unless the doc explicitly says so. If continuity is needed for the next op, the best thing is to add a `.contiguous()` on it.

**Comment by OverLordGoldDragon** [comment]
@SsnL Fair, agreed.
